File: packages/ui-swiftui/Sources/KinetixUI/Image.swift
//
// Image.swift — KinetixImage.
//
// Mirrors packages/ui/src/components/image.tsx: a ratio-locked image with
// a muted placeholder while loading and a fallback on error (presets 1:1
// / 3:2 / 4:3 / 3:4 / 3:1 / 16:9, or `.custom(w/h)`). Built on
// `AsyncImage` — its `.empty` / `.success` / `.failure` phases map
// directly onto the React `loading` / `loaded` / `error` states.
//

import SwiftUI

public enum KinetixImageRatio {
    case square, threeToTwo, fourToThree, threeToFour, threeToOne, sixteenToNine
    case custom(CGFloat)

    var value: CGFloat {
        switch self {
        case .square:        return 1
        case .threeToTwo:    return 3.0 / 2.0
        case .fourToThree:   return 4.0 / 3.0
        case .threeToFour:   return 3.0 / 4.0
        case .threeToOne:    return 3
        case .sixteenToNine: return 16.0 / 9.0
        case .custom(let v): return v
        }
    }
}

public struct KinetixImage: View {
    @Environment(\.kinetixColors) private var colors

    private let url: URL?
    private let ratio: KinetixImageRatio
    private let rounded: Bool

    public init(url: URL?, ratio: KinetixImageRatio = .square, rounded: Bool = true) {
        self.url = url
        self.ratio = ratio
        self.rounded = rounded
    }

    public var body: some View {
        Color.clear
            .aspectRatio(ratio.value, contentMode: .fit)
            .overlay {
                AsyncImage(url: url) { phase in
                    switch phase {
                    case .success(let image):
                        image.resizable().scaledToFill()
                    case .failure:
                        colors.muted.overlay {
                            Image(systemName: "photo")
                                .font(.system(size: 24))
                                .foregroundStyle(colors.mutedForeground)
                        }
                    case .empty:
                        colors.muted
                    @unknown default:
                        colors.muted
                    }
                }
            }
            .clipped()
            .clipShape(RoundedRectangle(cornerRadius: rounded ? 8 : 0, style: .continuous))
    }
}

File: packages/ui-flutter/lib/src/image.dart
import 'package:flutter/material.dart';

import 'theme.dart';

enum KinetixImageRatio { square, threeToTwo, fourToThree, threeToFour, threeToOne, sixteenToNine }

/// Mirrors `packages/ui/src/components/image.tsx`: a ratio-locked network
/// image with a muted placeholder while loading and a fallback on error
/// (presets 1:1 / 3:2 / 4:3 / 3:4 / 3:1 / 16:9, or `customRatio`).
class KinetixImage extends StatelessWidget {
  const KinetixImage({
    super.key,
    required this.url,
    this.ratio = KinetixImageRatio.square,
    this.rounded = true,
    this.customRatio,
  });

  final String url;
  final KinetixImageRatio ratio;
  final bool rounded;
  final double? customRatio;

  double get _ratio =>
      customRatio ??
      switch (ratio) {
        KinetixImageRatio.square => 1,
        KinetixImageRatio.threeToTwo => 3 / 2,
        KinetixImageRatio.fourToThree => 4 / 3,
        KinetixImageRatio.threeToFour => 3 / 4,
        KinetixImageRatio.threeToOne => 3,
        KinetixImageRatio.sixteenToNine => 16 / 9,
      };

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    return ClipRRect(
      borderRadius: BorderRadius.circular(rounded ? 8 : 0),
      child: AspectRatio(
        aspectRatio: _ratio,
        child: Image.network(
          url,
          fit: BoxFit.cover,
          loadingBuilder: (ctx, child, progress) => progress == null ? child : ColoredBox(color: c.muted),
          errorBuilder: (ctx, error, stackTrace) => ColoredBox(
            color: c.muted,
            child: Center(
              child: Icon(Icons.broken_image_outlined, size: 24, color: c.mutedForeground),
            ),
          ),
        ),
      ),
    );
  }
}
